src/domain/services/lottoService.js

```diff
@@ -0,0 +1,32 @@
+import CONFIG from '../../constants/config';
+
+const lottoService = {
+  calculateProfit(matchedResultList, purchaseAmount) {
+    const rankCounts = this.calculateRankCounts(matchedResultList);
+    const totalPrice = Array.from({ length: CONFIG.LOTTO_RANK_LENGTH }, (_, i) => i + 1).reduce(
+      (acc, rank) => acc + rankCounts[rank] * CONFIG.PRIZE[rank],
+      0,
+    );
+
+    return ((totalPrice / purchaseAmount) * 100).toFixed(CONFIG.PROFIT_DECIMAL_PLACE);
+  },
+
+  calculateRankCounts(matchedResultList) {
+    const rankCounts = { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0 };
+    matchedResultList.forEach(matchedResult => {
+      const rank = this.pickLottoRank(matchedResult);
+      rankCounts[rank] += 1;
+    });
```

Comment:
forEach를 그다지 싫어하는 것은 아니지만, reduce를 써도 되지 않았을까 싶긴 하네요!

Reply:
꽤 오래 고민을 해봤는데 reduce로 어떻게 구현할 수 있을지 감이 안잡히네요,,,calculateProfit 메서드에서는 totalPrice를 구할 때 reduce의 반환값이 number여서 적절히 사용할 수 있었습니다. 어떻게 접근해야 reduce로 사용할 수 있는지 궁금합니다! 
등수 카운트를 계산하는 로직은 다음과 같습니다.
- rankCounts : `{ 등수: 당첨 개수 }` 를 갖는 객체
- matchedResultList : `{ 당첨개수, 보너스 일치 여부 }` 인 matchedResult를 요소로 갖는 배열
- pickLottoRank에 matchedResult를 넘기면 `등수를 number 타입으로 반환`
- 등수를 키값으로 참조하여 rankCounts `갱신`
- matchedResultList를 순회하여 { 등수: 당첨 개수 } 결과를 `모두 갱신한 rankCounts 반환`

Reply:
reduce는 사실상 어떤 배열을 토대로 어떤 타입이든 하나의 결과물로 만드는 데에 모두 사용할 수 있다고 봅니다! 그게 숫자거나 문자열이거나 객체거나 상관 없이요! 이렇게 하면 가능하지 않을까요??

```suggestion
    const rankCounts = matchedResultList.reduce((ranks, matchedResult) => {
      const rank = this.pickLottoRank(matchedResult);
      ranks[rank] += 1;
      return ranks
    }, { 1: 0, 2: 0, 3: 0, 4: 0, 5: 0, 6: 0 });
```

Reply:
오호 이렇게 적용할 수 있군요! 처음봤을 때는 이와 같은 로직이라면 forEach와 크게 다르지 않다고 느껴지기도 합니다. 그런데 좀더 생각을 해봤을 때는 reduce를 제안해주실 때 다음과 같은 의미가 있었을 것 같아요.

하나의 객체(rankCounts)에 누적해서 더한다는 의미로 reduce가 가독성 측면에서 적합
기존의 rankCounts 객체에 접근하는 게 아니므로 초기값 {1:0, ... 6:0} 을 상수로 뺄 수 있음
기존의 객체를 참조하는 것이 아닌 새로운 객체를 만들어 불변성(?)을 지킴